de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SarosSession.java

```diff
@@ -1037,13 +1040,22 @@ public void changeColor(int colorID) {
   }
 
   @Override
-  public void enableQueuing(IProject project) {
-    activityQueuer.enableQueuing(project);
+  public void registerActivityQueuer(IActivityQueuer activityQueuer) {
+    if (activityQueuer == null || !activityQueuerChangeAllowed) {
+      return;
+    }
+
+    synchronized (this.activityQueuer) {
+      if (activityQueuerChangeAllowed) {
+        log.info("register a new Activity Queue Handler: " + activityQueuer);
+        this.activityQueuer = activityQueuer;
```

Comment:
You replace the `activityQueuer` object that you are using as a lock. The way I would understand locking, this would allow another thread to enter the synchronized block before the current thread has left it as you have replaced the object used for locking. Or am I wrong on this?

If this is the case, the current logic could lead to concurrency issues as you replace the lock-object before setting the if condition. This would allow another thread to enter the synchronized block after the `activityQueuer` was replaced but before `activityQueuerChangeAllowed` is set to false by the current thread, meaning the second thread could proceed into the if block and also replace the `activityQueuer` object.

To avoid this problem (if it is a problem), you have to set `activityQueuerChangeAllowed` before replacing the object or use an unrelated object as the lock.

Reply:
I do not think it has to be thread safe, as it will be not called concurrently anyway

What you can indeed do is use the default Queuer (I think you had one in the former patch set) and then do

registerActivityQeuer(IActivityQueuer queuer) if (queuer == null) throw ... if (currentQueuer != DEFAULT_QUEUER) throw ... currentQueuer = queuer

However as we already know it is set over and over again and the code now looks like this. And this is bad API design. 

You can use a static block inside the IPNs or add a custom component that registers the required upon session startup.

If you do not want to throw an exception return at least true or false so the caller could get a feedback.

And as Tobias already mentioned synchronizing on the passed in activityQueuer object does not make any sense.

Reply:
Yes sry this sync block was dump, I changed it to an AtomicBoolean and added a return if the call was successful.